### DISCUSSION TITLE
Data-dependent query refetch interval

### DISCUSSION THREAD

**Discussion by remko** [TanStack/query#2086] [Q&A]
I have a query that, depending on the data that is returned, needs to be refetched sooner than usual.
E.g., the data returned could contain a field 'stillComputing' to indicate that the result is not finished yet, and you should ask for the result again 'soon'. 

My initial intuition would be to be able to pass a function to a query's `refetchInterval` which takes the current data, and returns a number depending on the data (in the above case e.g. 5 seconds when the `stillComputing` field is set, or `undefined` otherwise), but according to the docs, that doesn't work.

The imperative way is to start a timer and call refetch() on the query (which isn't that bad if you have hooks to do this), but I was wondering if there was a declarative way.

**Comment by TkDodo** [comment]
local state combined with an updater from `onSuccess` is what you'll likely want for that:

```
const useMyQuery = () => {
  const [stillComputing, setStillComputing] = React.useState(false)
  return useQuery(
    'key',
    fetchFn,
    {
      refetchInterval: stillComputing ? undefined : 5000,
      onSuccess: (data) => {
        setStillComputing(data.stillComputing)
      }
    }
}
```
